### DISCUSSION TITLE
[FEATURE] Evaluate Profile-Guided Optimization (PGO) and LLVM BOLT

### DISCUSSION THREAD

**Discussion by zamazan4ik** [juspay/hyperswitch#2823] [Ideas / Feature Requests]
### Feature Description

Not a feature - just an idea of how to improve Hyperswitch performance (CPU usage part).

### Possible Implementation

Recently I did many Profile-Guided Optimization (PGO) benchmarks on multiple projects (including many network-related projects like Envoy, HAProxy, etc.) - the results are available [here](https://github.com/zamazan4ik/awesome-pgo). I think it could interesting to try to test PGO for optimizing the CPU part of Hyperswitch.

I can suggest the following things to do:
* Evaluate PGO's applicability to Hyperswitch.
* If PGO helps to achieve better performance - add a note to Hyperswitch's documentation about that. In this case, users and maintainers will be aware of another optimization opportunity for Hyperswitch.
* Provide PGO integration into the build scripts. It can help users and maintainers easily apply PGO for their own workloads.

Here are some examples of how PGO is already integrated into other projects' build scripts:
* Rustc: a CI [script](https://github.com/rust-lang/rust/blob/master/src/ci/stage-build.py) for the multi-stage build
* GCC:
  - Official [docs](https://gcc.gnu.org/install/build.html), section "Building with profile feedback" (even AutoFDO build is supported)
  - A [part](https://github.com/gcc-mirror/gcc/blob/4832767db7897be6fb5cbc44f079482c90cb95a6/configure#L7818) in a "wonderful" `configure` script 
* Clang: [Docs](https://llvm.org/docs/HowToBuildWithPGO.html) 
* Python: 
  - CPython: [README](https://github.com/python/cpython#profile-guided-optimization)
  - Pyston: [README](https://github.com/pyston/pyston#building)
* Go: [Bash script](https://github.com/golang/go/blob/master/src/cmd/compile/profile.sh)
* V8: [Bazel flag](https://github.com/v8/v8/blob/main/BUILD.gn#L184)
* ChakraCore: [Scripts](https://github.com/chakra-core/ChakraCore/tree/master/Build/scripts/pgo)
* Chromium: [Script](https://chromium.googlesource.com/chromium/src/build/config/+/refs/heads/main/compiler/pgo/BUILD.gn)
* Firefox: [Docs](https://firefox-source-docs.mozilla.org/build/buildsystem/pgo.html)
   - Thunderbird has PGO support too
* PHP - [Makefile command](https://github.com/php/php-src/blob/master/build/Makefile.global#L138) and old Centminmod [scripts](https://github.com/centminmod/php_pgo_training_scripts)
* MySQL: [CMake script](https://github.com/mysql/mysql-server/blob/8.0/cmake/fprofile.cmake)
* YugabyteDB: [GitHub commit](https://github.com/yugabyte/yugabyte-db/commit/34cb791ed9d3d5f8ae9a9b9e9181a46485e1981d)
* FoundationDB: [Script](https://github.com/apple/foundationdb/blob/1a6114a66f3de508c0cf0a45f72f3687ba05750c/contrib/generate_profile.sh)
* Zstd: [Makefile](https://github.com/facebook/zstd/blob/dev/programs/Makefile#L232)
* [Foot](https://codeberg.org/dnkl/foot): [Scripts](https://codeberg.org/dnkl/foot/src/branch/master/pgo)
* Windows Terminal: [GitHub PR](https://github.com/microsoft/terminal/pull/10071)
* Pydantic-core: [GitHub PR](https://github.com/pydantic/pydantic-core/pull/741)

After PGO, I can suggest evaluating [LLVM BOLT](https://github.com/llvm/llvm-project/blob/main/bolt/README.md) as an additional optimization step after PGO.

For the Rust projects, I recommend starting with [cargo-pgo](https://github.com/Kobzol/cargo-pgo).

### Have you spent some time checking if this feature request has been raised before?

- [X] I checked and didn't find a similar issue

### Have you read the Contributing Guidelines?

- [X] I have read the [Contributing Guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md)

### Are you willing to submit a PR?

No, I don't have time to work on this right now

**Comment by ashokkjag** [comment]
Hi @zamazan4ik , Thanks for the detailed issue with excellent resources for reference.

Since code path execution/profiling will need some external triggers/API calls, I'll leave this unassigned for now until someone can pick up and add it to appropriate CI.
